services/src/main/java/org/keycloak/services/managers/AuthenticationSessionManager.java

```diff
@@ -262,15 +263,22 @@ public boolean removeTabIdInAuthenticationSession(RealmModel realm, Authenticati
      * @param authSession
      */
     public void updateAuthenticationSessionAfterSuccessfulAuthentication(RealmModel realm, AuthenticationSessionModel authSession) {
-        // TODO: The authentication session might need to be expired in short interval (realm accessCodeLifespan, which is 1 minute by default). That should be sufficient for other browser tabs
-        //  to finish authentication and at the same time we won't need to keep authentication sessions in storage longer than needed
         boolean removedRootAuthSession = removeTabIdInAuthenticationSession(realm, authSession);
         if (!removedRootAuthSession) {
             RootAuthenticationSessionModel rootAuthSession = authSession.getParentSession();
 
-            log.tracef("Removed authentication session of root session '%s' with tabId '%s'. But there are remaining tabs in the root session", rootAuthSession.getId(), authSession.getTabId());
+            // 1 minute by default. Same timeout, which is used for client to complete "authorization code" flow
+            // Very short timeout should be OK as when this cookie is set, other existing browser tabs are supposed to be refreshed immediately by JS script authChecker.js
+            // and login user automatically. No need to have authenticationSession and cookie living any longer
+            int authSessionExpiresIn = realm.getAccessCodeLifespan();
 
-            AuthenticationStateCookie.generateAndSetCookie(session, realm, rootAuthSession);
+            // Set timestamp to the past to make sure that authSession is scheduled for expiration in "authSessionExpiresIn" seconds
+            int authSessionExpirationTime = Time.currentTime() - SessionExpiration.getAuthSessionLifespan(realm) + authSessionExpiresIn;
+            rootAuthSession.setTimestamp(authSessionExpirationTime);
```

Comment:
It took me some time understand that we are moving the timestamp to the past to expire it in `authSessionExpiresIn`. As the `RootAuthenticationSessionAdapter` is using the timestamp to calculate the expiration. Just a comment for other reviewers. :smile:

Reply:
I think, I don't like this kind of hacks, but given the fact we don't have expiration field for expirable entities as we had in new store, I don't see any other option. Just a thought for possible future refactoring.

Reply:
+1, expiration field would be nice in the future.